### PR TITLE
[Enhancement]substring support constant fold

### LINKED ISSUE #23708: substring constant folding
## Enhancement

For query like this:
```
select substring("abc", 1,2); 
```
The expr in this query will evaluated on BE, so the possibility of using predicates to prune partitions and buckets is lost, so we should evaluate this expr in FE(we call applying the function to constant arguments "function constant folding").

Some source code file involved:
```
fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java
fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/FoldConstantsRule.java
```

We can reference "date_trunc" function.

After substring constant folding is implemented, we can test this in two ways:

1. Use explain command to ensure that the function is evaluated in FE instead BE,  the function behavior in BE and FE is consistent.
2. `select * from t where dt = "2022-01-02" or dt = cast(substring("2022-01-03", 1, 10) as date)`  will be optimized into `select ... where dt in ("2022-01-02", "2022-01-03")`.

### PR DESCRIPTION
**Pull request by ucasfl**
## Problem Summary:
Fixes #23708 (issue)

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 7 / 7 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java|7|7|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java [hunk 1]
```diff
@@ -892,6 +892,27 @@ public static ConstantOperator version() {
         return ConstantOperator.createVarchar(Config.mysql_server_version);
     }
 
+    @ConstantFunction.List(list = {
```
Comment (satanson) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java:895]: "substring" has an alias "substr", could you add it?

File: fe/fe-core/src/test/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctionsTest.java [hunk 2]
```diff
@@ -1108,4 +1108,26 @@ public void testNowWithParameter() throws AnalysisException {
...
+        assertEquals("abcd", ScalarOperatorFunctions.substring(ConstantOperator.createVarchar("abcd"),
+                ConstantOperator.createInt(1), ConstantOperator.createInt(10)).getVarchar());
+        assertEquals("cd", ScalarOperatorFunctions.substring(ConstantOperator.createVarchar("abcd"),
+                ConstantOperator.createInt(3), ConstantOperator.createInt(4)).getVarchar());
+        assertEquals("", ScalarOperatorFunctions.substring(ConstantOperator.createVarchar("abcd"),
+                ConstantOperator.createInt(0), ConstantOperator.createInt(2)).getVarchar());
+        assertEquals("", ScalarOperatorFunctions.substring(ConstantOperator.createVarchar("abcd"),
+                ConstantOperator.createInt(5), ConstantOperator.createInt(2)).getVarchar());
+    }
+
```
Comment (satanson) [inline fe/fe-core/src/test/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctionsTest.java:1132]: could you add another test?
use UtFrameUtils.getVerboseFragmentPlan to get plan, then assert that the constant is fold 
```
select * from t where dt = "2022-01-02" or dt = cast(substring("2022-01-03", 1, 10) as date) will be optimized into select ... where dt in ("2022-01-02", "2022-01-03").
```

**Review by Seaven — DISMISSED** [review]

**Comment by ucasfl** [comment]
I have no idea why these fe test failed after rebase main.

**Review by Seaven — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 7 / 7 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java|7|7|100.00%|[]|

**Review by satanson — APPROVED** [review]

**Review by stdpain — APPROVED** [review]

**Comment by satanson** [comment]
@mergify backport branch-2.5

**Comment by satanson** [comment]
@mergify backport branch-3.0

**Pull request merged**